File: apps/api/src/routes/notifications.ts
/** Notification routes (SRS §6.7, §7.7). notification.send. */
import { Hono } from "hono";
import { createNotificationSchema } from "@tuition/shared";
import type { AppContext, Db } from "../types";
import { parseBody } from "../lib/validate";
import { newId, nowIso } from "../lib/id";
import { writeAudit } from "../lib/db";
import { authenticate, requirePermission } from "../middleware/auth";

const COLS = `id, type, title, body, channel, audience, class_id, student_id,
  recipient_count, status, scheduled_at, sent_at, created_at`;

/** Count recipients for an audience (informational; delivery lands in M5). */
async function recipientCount(
  db: Db,
  audience: string,
  classId?: string | null,
  studentId?: string | null,
): Promise<number> {
  if (audience === "all_students") {
    const r = await db.prepare(`SELECT COUNT(*) AS n FROM students WHERE status = 'active' AND deleted_at IS NULL`).first<{ n: number }>();
    return r?.n ?? 0;
  }
  if (audience === "all_guardians") {
    const r = await db.prepare(`SELECT COUNT(*) AS n FROM guardians`).first<{ n: number }>();
    return r?.n ?? 0;
  }
  if (audience === "class" && classId) {
    const r = await db.prepare(`SELECT COUNT(*) AS n FROM enrollments WHERE class_id = ? AND status = 'active'`).bind(classId).first<{ n: number }>();
    return r?.n ?? 0;
  }
  if (audience === "student") return 1;
  return 0;
}

export const notificationsRoutes = new Hono<AppContext>();
notificationsRoutes.use("*", authenticate);

notificationsRoutes.get("/", requirePermission("notification.send"), async (c) => {
  const rows = await c
    .get("db")
    .prepare(`SELECT ${COLS} FROM notifications ORDER BY created_at DESC LIMIT 50`)
    .all();
  return c.json({ notifications: rows.results ?? [] });
});

notificationsRoutes.post("/", requirePermission("notification.send"), async (c) => {
  const body = await parseBody(c, createNotificationSchema);
  const db = c.get("db");
  const now = nowIso();
  const id = newId("ntf");
  const count = await recipientCount(db, body.audience, body.class_id, body.student_id);

  // Future schedule → queued; otherwise recorded as sent now (no provider yet).
  const scheduled = body.scheduled_at && body.scheduled_at > now ? body.scheduled_at : null;
  const status = scheduled ? "queued" : "sent";
  const sentAt = scheduled ? null : now;

  await db
    .prepare(
      `INSERT INTO notifications
         (id, type, title, body, channel, audience, class_id, student_id, recipient_count, status, scheduled_at, sent_at, created_by, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    )
    .bind(
      id, body.type, body.title, body.body, body.channel, body.audience,
      body.class_id ?? null, body.student_id ?? null, count, status, scheduled, sentAt, c.get("user").id, now, now,
    )
    .run();
  await writeAudit(db, { actorId: c.get("user").id, action: "notification.send", entityType: "notification", entityId: id, after: { audience: body.audience, status, recipient_count: count } });

  const row = await db.prepare(`SELECT ${COLS} FROM notifications WHERE id = ?`).bind(id).first();
  return c.json(row, 201);
});

File: apps/api/internal/notifications/notifications.go
// Package notifications serves notification list/create (notification.send).
// Delivery providers land later; this records intent + recipient counts.
package notifications

import (
	"context"
	"database/sql"
	"net/http"

	"github.com/go-chi/chi/v5"

	"attendly/api/internal/auth"
	"attendly/api/internal/cryptox"
	"attendly/api/internal/httpapi"
	"attendly/api/internal/store"
)

const cols = `id, type, title, body, channel, audience, class_id, student_id,
	recipient_count, status, scheduled_at, sent_at, created_at`

// Handlers serves the notifications domain.
type Handlers struct{ db *sql.DB }

// New constructs the notifications handlers.
func New(db *sql.DB) *Handlers { return &Handlers{db: db} }

// Mount registers /api/notifications (notification.send).
func (h *Handlers) Mount(r chi.Router) {
	send := auth.RequirePermission("notification.send")
	r.Route("/api/notifications", func(r chi.Router) {
		r.With(send).Method(http.MethodGet, "/", httpapi.Handler(h.list))
		r.With(send).Method(http.MethodPost, "/", httpapi.Handler(h.create))
	})
}

func (h *Handlers) recipientCount(ctx context.Context, audience string, classID, studentID *string) (int64, error) {
	var n int64
	switch {
	case audience == "all_students":
		return n, h.db.QueryRowContext(ctx, `SELECT COUNT(*) FROM students WHERE status = 'active' AND deleted_at IS NULL`).Scan(&n)
	case audience == "all_guardians":
		return n, h.db.QueryRowContext(ctx, `SELECT COUNT(*) FROM guardians`).Scan(&n)
	case audience == "class" && classID != nil:
		return n, h.db.QueryRowContext(ctx, `SELECT COUNT(*) FROM enrollments WHERE class_id = ? AND status = 'active'`, *classID).Scan(&n)
	case audience == "student":
		return 1, nil
	}
	return 0, nil
}

func (h *Handlers) list(w http.ResponseWriter, r *http.Request) error {
	rows, err := store.QueryMaps(r.Context(), h.db, `SELECT `+cols+` FROM notifications ORDER BY created_at DESC LIMIT 50`)
	if err != nil {
		return err
	}
	httpapi.JSON(w, http.StatusOK, map[string]any{"notifications": rows})
	return nil
}

func (h *Handlers) create(w http.ResponseWriter, r *http.Request) error {
	var in struct {
		Type        string  `json:"type"`
		Title       string  `json:"title"`
		Body        string  `json:"body"`
		Channel     string  `json:"channel"`
		Audience    string  `json:"audience"`
		ClassID     *string `json:"class_id"`
		StudentID   *string `json:"student_id"`
		ScheduledAt *string `json:"scheduled_at"`
	}
	if err := httpapi.Decode(r, &in); err != nil {
		return err
	}
	if in.Type == "" || in.Title == "" || in.Body == "" || in.Channel == "" || in.Audience == "" {
		return httpapi.BadRequest("invalid_body")
	}
	now := cryptox.NowISO()
	id := cryptox.NewID("ntf")
	count, err := h.recipientCount(r.Context(), in.Audience, in.ClassID, in.StudentID)
	if err != nil {
		return err
	}

	var scheduled, sentAt *string
	status := "sent"
	if in.ScheduledAt != nil && *in.ScheduledAt > now {
		scheduled = in.ScheduledAt
		status = "queued"
	} else {
		sentAt = &now
	}

	actor := auth.MustUser(r.Context()).ID
	if _, err := h.db.ExecContext(r.Context(),
		`INSERT INTO notifications
		   (id, type, title, body, channel, audience, class_id, student_id, recipient_count, status, scheduled_at, sent_at, created_by, created_at, updated_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		id, in.Type, in.Title, in.Body, in.Channel, in.Audience, in.ClassID, in.StudentID, count, status, scheduled, sentAt, actor, now, now); err != nil {
		return err
	}
	_ = store.WriteAudit(r.Context(), h.db, store.AuditEntry{ActorID: &actor, Action: "notification.send", EntityType: sp("notification"), EntityID: &id, After: map[string]any{"audience": in.Audience, "status": status, "recipient_count": count}})

	row, err := store.QueryFirstMap(r.Context(), h.db, `SELECT `+cols+` FROM notifications WHERE id = ?`, id)
	if err != nil {
		return err
	}
	httpapi.JSON(w, http.StatusCreated, row)
	return nil
}

func sp(s string) *string { return &s }
